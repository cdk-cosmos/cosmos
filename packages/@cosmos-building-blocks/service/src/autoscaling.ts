import { Construct, Lazy, Fn } from '@aws-cdk/core';
import { InstanceType, InstanceClass, InstanceSize, CfnLaunchTemplate } from '@aws-cdk/aws-ec2';
import { CfnInstanceProfile } from '@aws-cdk/aws-iam';
import {
  CfnAutoScalingGroup,
  AutoScalingGroup,
  AutoScalingGroupProps,
  Monitoring,
  BlockDevice,
  EbsDeviceVolumeType,
  BlockDeviceVolume,
} from '@aws-cdk/aws-autoscaling';

/**
 * Properties for AutoScalingGroupLT
 */
export interface AutoScalingGroupLTProps extends AutoScalingGroupProps {
  /**
   * Set this to `true` if you want to use Instance Template instead of Launch Config.
   * You can fine tune your default LaunchTemplate with `instanceType`, `launchTemplateOverrides` and `instancesDistribution`
   * @default - false
   */
  readonly useInstanceTemplate?: boolean;
  /**
   * Type of instance to launch. if used with `useInstanceTemplate = true`, it will be applied to LaunchTemplate.
   * Make sure to check and tweak `launchTemplateOverrides` and `instancesDistribution` as the default values set for those
   * properties will overwrite this InstanceType
   */
  readonly instanceType: InstanceType;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Array of LaunchTemplateOverridesProperty which is a subproperty of LaunchTemplate that describes an override for a launch template.
   * Currently, the only supported override is instance type
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html
   * @default
   * [{ instanceType: "t3.medium" }, { instanceType: "t3.small" }]
   *
   * This will use any Spot instances available, "t3.medium" or "t3.small"
   * based on `spotAllocationStrategy` set in `instancesDistribution`
   */
  readonly launchTemplateOverrides?: Array<CfnAutoScalingGroup.LaunchTemplateOverridesProperty>;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Specifies the distribution of On-Demand Instances and Spot Instances,
   * the maximum price to pay for Spot Instances, and how the Auto Scaling group allocates instance types
   * to fulfill On-Demand and Spot capacity.
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-instancesdistribution.html
   * @default - Uses all Spot instances of types specified in property `launchTemplateOverrides`
   *  {
   *    onDemandAllocationStrategy: "prioritized",
   *    onDemandBaseCapacity: 0,
   *    onDemandPercentageAboveBaseCapacity: 0,
   *    spotAllocationStrategy: "capacity-optimized",
   *    spotMaxPrice: props.spotPrice,
   *  }
   */
  readonly instancesDistribution?: Partial<CfnAutoScalingGroup.InstancesDistributionProperty>;
}

/**
 * Derived from Base class AutoScalingGroup with additional functionalities of using Launch Template (LT).
 *
 * Base AutoScalingGroup class uses Launch Config and have no ability to use more advanced Launch Templates
 * and MixedInstancesPolicies which gives ability to use Spot fleet and combination of spot and on-demand
 *
 * This construct uses default spot fleet template settings with just one property `useInstanceTemplate = true`
 *
 * With `launchTemplateOverrides` and `instancesDistribution`, it gives additional fine tuning.
 */
export class AutoScalingGroupLT extends AutoScalingGroup {
  constructor(scope: Construct, id: string, props: AutoScalingGroupLTProps) {
    super(scope, id, props);

    // Check if useInstanceTemplate is set to true. If not, do nothing and rely on Base class to
    // to create AutoScalingGroup using LaunchConfig
    const useInstanceTemplate = props.useInstanceTemplate !== undefined ? props.useInstanceTemplate === true : false;
    if (useInstanceTemplate == true) {
      // Prepare props required for LaunchTemplate
      const instanceType =
        props.instanceType !== undefined
          ? props.instanceType.toString()
          : InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString();

      const imageConfig = props.machineImage.getImage(this);
      const userDataToken = Lazy.stringValue({
        produce: () => Fn.base64(this.userData.render()),
      });
      const securityGroupIds = this.connections.securityGroups.map(sg => sg.securityGroupId);
      const instanceMonitoringMode =
        props.instanceMonitoring !== undefined ? props.instanceMonitoring === Monitoring.DETAILED : undefined;
      const monitoring = {
        enabled: instanceMonitoringMode == undefined ? true : instanceMonitoringMode,
      };

      const iamProfile = this.node.tryFindChild('InstanceProfile') as CfnInstanceProfile;

      // Create LaunchTemplate

      const lTemplate = new CfnLaunchTemplate(this, 'LaunchTemplate', {
        launchTemplateData: {
          imageId: imageConfig.imageId,
          securityGroupIds,
          userData: userDataToken,
          iamInstanceProfile: { arn: iamProfile.attrArn },
          keyName: props.keyName,
          monitoring,
          instanceType: instanceType,
          blockDeviceMappings:
            props.blockDevices !== undefined ? synthesizeBlockDeviceMappings(this, props.blockDevices) : undefined,
        },
      });
      lTemplate.node.addDependency(this.role);

      // Prepare props for ASG with MixedInstancesPolicy

      const overrides =
        props.launchTemplateOverrides == undefined
          ? [
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM).toString(),
              },
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString(),
              },
            ]
          : props.launchTemplateOverrides;

      const instancesDistribution = {
        onDemandAllocationStrategy: 'prioritized',
        onDemandBaseCapacity: 0,
        onDemandPercentageAboveBaseCapacity: 0,
        spotAllocationStrategy: 'capacity-optimized',
        spotMaxPrice: props.spotPrice,
        ...props.instancesDistribution,
      };

      // Mixed instance Policy
      const mixedInstancesPolicy: CfnAutoScalingGroup.MixedInstancesPolicyProperty = {
        launchTemplate: {
          launchTemplateSpecification: {
            version: lTemplate.attrLatestVersionNumber,
            launchTemplateId: lTemplate.ref,
          },
          overrides,
        },
        instancesDistribution,
      };

      // Remove LaunchConfig from the resulting CloudFormation template
      if (!this.node.tryRemoveChild('LaunchConfig')) throw new Error('Launch config not found');
      const asg = this.node.tryFindChild('ASG') as CfnAutoScalingGroup;
      asg.addPropertyDeletionOverride('LaunchConfigurationName');
      // Inject mixedInstancesPolicy to use LaunchTemplate
      asg.mixedInstancesPolicy = mixedInstancesPolicy;
    }
  }
}
function synthesizeBlockDeviceMappings(
  construct: Construct,
  blockDevices: BlockDevice[]
): CfnLaunchTemplate.BlockDeviceMappingProperty[] {
  return blockDevices.map<CfnLaunchTemplate.BlockDeviceMappingProperty>(({ deviceName, volume, mappingEnabled }) => {
    const { virtualName, ebsDevice: ebs } = volume;

    if (volume === BlockDeviceVolume.noDevice() || mappingEnabled === false) {
      return {
        deviceName,
        noDevice: 'true',
      };
    }

    if (ebs) {
      const { iops, volumeType } = ebs;

      if (!iops) {
        if (volumeType === EbsDeviceVolumeType.IO1) {
          throw new Error('iops property is required with volumeType: EbsDeviceVolumeType.IO1');
        }
      } else if (volumeType !== EbsDeviceVolumeType.IO1) {
        construct.node.addWarning('iops will be ignored without volumeType: EbsDeviceVolumeType.IO1');
      }
    }

    return {
      deviceName,
      ebs,
      virtualName,
    };
  });
}
