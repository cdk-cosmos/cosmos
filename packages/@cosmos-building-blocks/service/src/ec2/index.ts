export {
  CloudWatchAgent,
  CloudWatchAgentProps,
  CloudWatchMetricBuilder,
  CloudWatchMetric,
} from './ec2-cloud-watch-agent';
