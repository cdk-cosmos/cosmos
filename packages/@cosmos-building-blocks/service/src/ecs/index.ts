export { EcsService, EcsServiceProps } from './ecs-service';
