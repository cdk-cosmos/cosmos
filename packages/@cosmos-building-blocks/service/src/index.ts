export { EcsService, EcsServiceProps } from './ecs';
