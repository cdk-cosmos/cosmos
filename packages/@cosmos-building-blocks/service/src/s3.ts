import { Construct } from '@aws-cdk/core';
import { Bucket, BucketEncryption, BucketProps, BucketAccessControl } from '@aws-cdk/aws-s3';
import { PolicyStatement, Effect, AnyPrincipal } from '@aws-cdk/aws-iam';

export class SecureBucket extends Bucket {
  constructor(scope: Construct, id: string, props?: BucketProps) {
    super(scope, id, {
      encryption: BucketEncryption.S3_MANAGED,
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      serverAccessLogsPrefix: 'BucketAccessLogs/',
      ...props,
    });
    const policyStatement = new PolicyStatement({
      actions: ['s3:*'],
      principals: [new AnyPrincipal()],
      resources: [`${this.bucketArn}/*`],
      effect: Effect.DENY,
      sid: 'ForceSSLOnly',
    });
    policyStatement.addCondition('Bool', { 'aws:SecureTransport': false });
    super.addToResourcePolicy(policyStatement);
  }
}
