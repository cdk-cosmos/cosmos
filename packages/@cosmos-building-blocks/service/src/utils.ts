export const getRoutingPriority = (props: { hostHeader?: string; pathPattern?: string }): number => {
  const path = getStringValue(props.pathPattern);
  const host = getStringValue(props.hostHeader);
  const count = 2;
  const total = path + host;
  const inMin = 0 * count;
  const inMax = 128 * charMapping.length * count;
  const outMin = 1000; // Offset
  const outMax = Math.min(inMax + outMin, 50000); // only use and many as needed ?

  // Since lower numbers are more important in Listener Rules for ALB, then flip out min and max
  return mapNumber(total, inMin, inMax, outMax, outMin);
};

const mapNumber = (value: number, inMin: number, inMax: number, outMin: number, outMax: number): number => {
  return Math.floor(((value - inMin) / (inMax - inMin)) * (outMax - outMin) + outMin);
};

const getStringValue = (str?: string): number => {
  if (str === undefined) return 0;
  let total = 0;
  for (let i = 0; i <= str.length; i++) {
    total += charMapping.indexOf(str.charAt(i)) + 1;
  }
  return total;
};

const charMapping: string[] = [
  '*',
  '?',
  '&',
  '_',
  '@',
  '+',
  "'",
  '"',
  '~',
  '$',
  '.',
  ':',
  '-',
  'a',
  'b',
  'c',
  'd',
  'e',
  'f',
  'g',
  'h',
  'i',
  'j',
  'k',
  'l',
  'm',
  'n',
  'o',
  'p',
  'q',
  'r',
  's',
  't',
  'u',
  'v',
  'w',
  'x',
  'y',
  'z',
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
  '0',
  '1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '/',
];
