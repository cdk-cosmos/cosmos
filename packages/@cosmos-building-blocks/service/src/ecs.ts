import { Construct, Duration } from '@aws-cdk/core';
import {
  Ec2TaskDefinition,
  Ec2Service,
  Protocol,
  ContainerDefinitionOptions,
  PortMapping,
  Ec2ServiceProps,
  ICluster,
  LogDrivers,
  ContainerDefinition,
  ScalableTaskCount,
  CpuUtilizationScalingProps,
  RequestCountScalingProps,
  PlacementStrategy,
  BuiltInAttributes,
  Ec2TaskDefinitionProps,
} from '@aws-cdk/aws-ecs';
import {
  ApplicationTargetGroup,
  ApplicationProtocol,
  ApplicationListenerRule,
  IApplicationListener,
  ApplicationListenerRuleProps,
  ApplicationTargetGroupProps,
} from '@aws-cdk/aws-elasticloadbalancingv2';
import { IVpc } from '@aws-cdk/aws-ec2';
import { LogGroup, LogGroupProps } from '@aws-cdk/aws-logs';
import { EnableScalingProps } from '@aws-cdk/aws-applicationautoscaling';
import { getRoutingPriority } from './utils';

export interface EcsServiceProps {
  vpc: IVpc;
  cluster: ICluster;
  containerProps: ContainerDefinitionOptions & { port?: PortMapping };
  httpListener?: IApplicationListener;
  taskProps?: Partial<Ec2TaskDefinitionProps>;
  serviceProps?: Partial<Ec2ServiceProps>;
  targetGroupProps?: Partial<ApplicationTargetGroupProps>;
  routingProps?: Partial<ApplicationListenerRuleProps>;
  scalingProps?: EnableScalingProps;
  logProps?: Partial<LogGroupProps>;
}

export class EcsService extends Construct {
  readonly LogGroup: LogGroup;
  readonly TaskDefinition: Ec2TaskDefinition;
  readonly Container: ContainerDefinition;
  readonly Service: Ec2Service;
  readonly TargetGroup?: ApplicationTargetGroup;
  readonly ListenerRule?: ApplicationListenerRule;
  readonly Scaling?: ScalableTaskCount;

  constructor(scope: Construct, id: string, props: EcsServiceProps) {
    super(scope, id);

    const {
      vpc,
      cluster,
      httpListener,
      containerProps,
      taskProps,
      serviceProps,
      targetGroupProps,
      routingProps,
      scalingProps,
      logProps,
    } = props;

    this.LogGroup = new LogGroup(this, 'Logs', logProps);

    this.TaskDefinition = new Ec2TaskDefinition(this, 'Task', {
      ...taskProps,
    });

    this.Container = this.TaskDefinition.addContainer('Container', {
      memoryLimitMiB: 256,
      logging: LogDrivers.awsLogs({ logGroup: this.LogGroup, streamPrefix: `Container` }),
      ...containerProps,
    });

    this.Container.addPortMappings({
      containerPort: 80,
      protocol: Protocol.TCP,
      ...containerProps.port,
    });

    this.Service = new Ec2Service(this, 'Service', {
      desiredCount: 1,
      placementStrategies: [PlacementStrategy.spreadAcross(BuiltInAttributes.AVAILABILITY_ZONE)],
      ...serviceProps,
      taskDefinition: this.TaskDefinition,
      cluster: cluster,
    });

    if (routingProps) {
      if (!httpListener) throw new Error('To enable routing, Http Listener is required');

      this.TargetGroup = new ApplicationTargetGroup(this, 'ServiceTargetGroup', {
        protocol: ApplicationProtocol.HTTP,
        deregistrationDelay: Duration.seconds(0),
        ...targetGroupProps,
        vpc: vpc,
        targets: [
          this.Service.loadBalancerTarget({
            containerName: 'Container',
          }),
        ],
      });

      this.ListenerRule = new ApplicationListenerRule(this, 'ServiceRule', {
        priority: getRoutingPriority(routingProps),
        ...routingProps,
        listener: httpListener,
        targetGroups: [this.TargetGroup],
      });
    }

    if (scalingProps) {
      this.Scaling = this.Service.autoScaleTaskCount(scalingProps);
    }
  }

  addCpuAutoScaling(props: Partial<CpuUtilizationScalingProps>): void {
    if (!this.Scaling) throw new Error('Scaling needs to be enabled');
    this.Scaling.scaleOnCpuUtilization('CpuScaling', {
      ...props,
      targetUtilizationPercent: 50,
    });
  }

  addMemoryAutoScaling(props: Partial<RequestCountScalingProps>): void {
    if (!this.Scaling) throw new Error('Scaling needs to be enabled');
    this.Scaling.scaleOnMemoryUtilization('MemoryScaling', {
      ...props,
      targetUtilizationPercent: 50,
    });
  }

  addRequestAutoScaling(props: Partial<RequestCountScalingProps>): void {
    if (!this.Scaling) throw new Error('Scaling needs to be enabled');
    if (!this.TargetGroup) throw new Error('Routing needs to be enabled');
    this.Scaling.scaleOnRequestCount('RequestScaling', {
      ...props,
      requestsPerTarget: 500,
      targetGroup: this.TargetGroup,
    });
  }
}
