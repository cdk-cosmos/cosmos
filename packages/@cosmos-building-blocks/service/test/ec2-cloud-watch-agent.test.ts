import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { App, Stack } from '@aws-cdk/core';
import { AutoScalingGroup } from '@aws-cdk/aws-autoscaling';
import { Vpc, Instance, InstanceType, MachineImage } from '@aws-cdk/aws-ec2';
import { CloudWatchAgent } from '../src';

const app = new App();
const stack = new Stack(app, 'Stack', { env: { region: 'region' } });
const instance = new Instance(stack, 'Instance', {
  vpc: Vpc.fromVpcAttributes(stack, 'Vpc', {
    vpcId: 'Vpc',
    availabilityZones: ['a'],
    isolatedSubnetIds: ['a-sub-1'],
  }),
  instanceType: new InstanceType('t3.nano'),
  machineImage: MachineImage.latestAmazonLinux(),
});
new CloudWatchAgent(stack, 'CloudWatchAgent', {
  compute: instance,
});

const synth = SynthUtils.toCloudFormation(stack);

test('Cloud Watch Agent', () => {
  const userData = stack.resolve(instance.userData.render());
  expect(userData).toContain('rpm');
  expect(userData).toContain('echo');
  expect(userData).toContain('amazon-cloudwatch-agent-ctl');
  expect(userData).toContain('mem_used_percent');
  expect(userData).toContain('used_percent');
  expect(userData).toMatchSnapshot();
});

test('Cloud Watch Agent Snapshot', () => {
  expect(synth).toMatchSnapshot();
});

test('Cloud Watch Agent with ASG', () => {
  const app = new App();
  const stack = new Stack(app, 'Stack', { env: { region: 'region' } });
  const asg = new AutoScalingGroup(stack, 'Instance', {
    vpc: Vpc.fromVpcAttributes(stack, 'Vpc', {
      vpcId: 'Vpc',
      availabilityZones: ['a'],
      isolatedSubnetIds: ['a-sub-1'],
    }),
    instanceType: new InstanceType('t3.nano'),
    machineImage: MachineImage.latestAmazonLinux(),
  });
  new CloudWatchAgent(stack, 'CloudWatchAgent', {
    compute: asg,
  });

  const synth = SynthUtils.toCloudFormation(stack);

  expect(synth).toMatchSnapshot();
});
