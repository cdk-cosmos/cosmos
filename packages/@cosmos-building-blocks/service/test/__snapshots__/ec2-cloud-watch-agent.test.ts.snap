// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Watch Agent 1`] = `
"#!/bin/bash
rpm -Uvh https://s3.region.amazonaws.com/amazoncloudwatch-agent-region/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm
echo -e \\"{\\\\\\"agent\\\\\\":{\\\\\\"metrics_collection_interval\\\\\\":60,\\\\\\"omit_hostname\\\\\\":true,\\\\\\"run_as_user\\\\\\":\\\\\\"root\\\\\\"},\\\\\\"metrics\\\\\\":{\\\\\\"append_dimensions\\\\\\":{\\\\\\"InstanceId\\\\\\":\\\\\\"\\\\\${aws:InstanceId}\\\\\\",\\\\\\"InstanceType\\\\\\":\\\\\\"\\\\\${aws:InstanceType}\\\\\\"},\\\\\\"metrics_collected\\\\\\":{\\\\\\"mem\\\\\\":{\\\\\\"namespace\\\\\\":\\\\\\"CWAgent\\\\\\",\\\\\\"interval\\\\\\":60,\\\\\\"measurement\\\\\\":[\\\\\\"mem_used_percent\\\\\\"]},\\\\\\"disk\\\\\\":{\\\\\\"namespace\\\\\\":\\\\\\"CWAgent\\\\\\",\\\\\\"interval\\\\\\":60,\\\\\\"measurement\\\\\\":[\\\\\\"used_percent\\\\\\"],\\\\\\"resources\\\\\\":[\\\\\\"*\\\\\\"]}}}}\\" > /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json
sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m ec2 -s -c file:/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json"
`;

exports[`Cloud Watch Agent Snapshot 1`] = `
Object {
  "Parameters": Object {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "InstanceC1063A87": Object {
      "DependsOn": Array [
        "InstanceInstanceRoleE9785DE5",
      ],
      "Properties": Object {
        "AvailabilityZone": "a",
        "IamInstanceProfile": Object {
          "Ref": "InstanceInstanceProfileAB5AEF02",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "InstanceInstanceSecurityGroupF0E2D5BE",
              "GroupId",
            ],
          },
        ],
        "SubnetId": "a-sub-1",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Instance",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash
rpm -Uvh https://s3.region.amazonaws.com/amazoncloudwatch-agent-region/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm
echo -e \\"{\\\\\\"agent\\\\\\":{\\\\\\"metrics_collection_interval\\\\\\":60,\\\\\\"omit_hostname\\\\\\":true,\\\\\\"run_as_user\\\\\\":\\\\\\"root\\\\\\"},\\\\\\"metrics\\\\\\":{\\\\\\"append_dimensions\\\\\\":{\\\\\\"InstanceId\\\\\\":\\\\\\"\\\\\${aws:InstanceId}\\\\\\",\\\\\\"InstanceType\\\\\\":\\\\\\"\\\\\${aws:InstanceType}\\\\\\"},\\\\\\"metrics_collected\\\\\\":{\\\\\\"mem\\\\\\":{\\\\\\"namespace\\\\\\":\\\\\\"CWAgent\\\\\\",\\\\\\"interval\\\\\\":60,\\\\\\"measurement\\\\\\":[\\\\\\"mem_used_percent\\\\\\"]},\\\\\\"disk\\\\\\":{\\\\\\"namespace\\\\\\":\\\\\\"CWAgent\\\\\\",\\\\\\"interval\\\\\\":60,\\\\\\"measurement\\\\\\":[\\\\\\"used_percent\\\\\\"],\\\\\\"resources\\\\\\":[\\\\\\"*\\\\\\"]}}}}\\" > /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json
sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m ec2 -s -c file:/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "InstanceInstanceProfileAB5AEF02": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "InstanceInstanceRoleE9785DE5",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "InstanceInstanceRoleE9785DE5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchAgentServerPolicy",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Instance",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "InstanceInstanceSecurityGroupF0E2D5BE": Object {
      "Properties": Object {
        "GroupDescription": "Stack/Instance/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Instance",
          },
        ],
        "VpcId": "Vpc",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
}
`;

exports[`Cloud Watch Agent with ASG 1`] = `
Object {
  "Parameters": Object {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "InstanceASG27CC400A": Object {
      "Properties": Object {
        "LaunchConfigurationName": Object {
          "Ref": "InstanceLaunchConfig62D85C8D",
        },
        "MaxSize": "1",
        "MinSize": "1",
        "Tags": Array [
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "Stack/Instance",
          },
        ],
        "VPCZoneIdentifier": Array [
          "a-sub-1",
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": Object {
        "AutoScalingScheduledAction": Object {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "InstanceInstanceProfileAB5AEF02": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "InstanceInstanceRoleE9785DE5",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "InstanceInstanceRoleE9785DE5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchAgentServerPolicy",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Instance",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "InstanceInstanceSecurityGroupF0E2D5BE": Object {
      "Properties": Object {
        "GroupDescription": "Stack/Instance/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Stack/Instance",
          },
        ],
        "VpcId": "Vpc",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "InstanceLaunchConfig62D85C8D": Object {
      "DependsOn": Array [
        "InstanceInstanceRoleE9785DE5",
      ],
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "InstanceInstanceProfileAB5AEF02",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "InstanceInstanceSecurityGroupF0E2D5BE",
              "GroupId",
            ],
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash
rpm -Uvh https://s3.region.amazonaws.com/amazoncloudwatch-agent-region/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm
echo -e \\"{\\\\\\"agent\\\\\\":{\\\\\\"metrics_collection_interval\\\\\\":60,\\\\\\"omit_hostname\\\\\\":true,\\\\\\"run_as_user\\\\\\":\\\\\\"root\\\\\\"},\\\\\\"metrics\\\\\\":{\\\\\\"append_dimensions\\\\\\":{\\\\\\"InstanceId\\\\\\":\\\\\\"\\\\\${aws:InstanceId}\\\\\\",\\\\\\"InstanceType\\\\\\":\\\\\\"\\\\\${aws:InstanceType}\\\\\\"},\\\\\\"metrics_collected\\\\\\":{\\\\\\"mem\\\\\\":{\\\\\\"namespace\\\\\\":\\\\\\"CWAgent\\\\\\",\\\\\\"interval\\\\\\":60,\\\\\\"measurement\\\\\\":[\\\\\\"mem_used_percent\\\\\\"]},\\\\\\"disk\\\\\\":{\\\\\\"namespace\\\\\\":\\\\\\"CWAgent\\\\\\",\\\\\\"interval\\\\\\":60,\\\\\\"measurement\\\\\\":[\\\\\\"used_percent\\\\\\"],\\\\\\"resources\\\\\\":[\\\\\\"*\\\\\\"]}}}}\\" > /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json
sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m ec2 -s -c file:/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json",
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
  },
}
`;
